src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -156,19 +156,18 @@ void Parse::array_store_check() {
   int klass_offset = oopDesc::klass_offset_in_bytes();
   Node* p = basic_plus_adr( ary, ary, klass_offset );
   // p's type is array-of-OOPS plus klass_offset
-  Node* array_klass = _gvn.transform(LoadKlassNode::make(_gvn, nullptr, immutable_memory(), p, TypeInstPtr::KLASS));
+  Node* array_klass = _gvn.transform(LoadKlassNode::make(_gvn, immutable_memory(), p, TypeInstPtr::KLASS));
   // Get the array klass
   const TypeKlassPtr *tak = _gvn.type(array_klass)->is_klassptr();
 
   // The type of array_klass is usually INexact array-of-oop.  Heroically
   // cast array_klass to EXACT array and uncommon-trap if the cast fails.
   // Make constant out of the inexact array klass, but use it only if the cast
   // succeeds.
-  bool always_see_exact_class = false;
-  if (MonomorphicArrayCheck
-      && !too_many_traps(Deoptimization::Reason_array_check)
-      && !tak->klass_is_exact()
-      && tak != TypeInstKlassPtr::OBJECT) {
+  if (MonomorphicArrayCheck &&
+      !too_many_traps(Deoptimization::Reason_array_check) &&
+      !tak->klass_is_exact() &&
+      tak->isa_aryklassptr()) {
```

Comment:
Looks like an implicit `nullptr` check. Not allowed by code style ;)

Reply:
Can you quickly explain this change from `tak != TypeInstKlassPtr::OBJECT` so I don't need to investigate myself, please?

Reply:
> Looks like an implicit nullptr check. Not allowed by code style ;)

But the verb here is `isa` and we use these as a `bool` a lot, though :/

> Can you quickly explain this change from tak != TypeInstKlassPtr::OBJECT so I don't need to investigate myself, please?

The bottom type of an array can be either `Object` or an array of some kind, so `tak != TypeInstKlassPtr::OBJECT` is the same as `tak->isa_aryklassptr()`.

Reply:
Ah great, thanks for the explanation!